text/0000-allow-if-let-irrefutables.md

```diff
@@ -0,0 +1,77 @@
+- Feature Name: allow_if_let_irrefutables
+- Start Date: 2017-07-27
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Currently when using an if let statement and an irrefutable pattern (read always match) is used the compiler complains with an `E0162: irrefutable if-let pattern`.
+The current state breaks macros who want to accept patterns generically and this RFC proposes changing this error to an error-by-default which is allowed to be disabled by such macros.
+
+# Motivation
+[motivation]: #motivation
+
+The use cases for this is in the creation of macros where patterns are allowed because to support the `_` patterns the code has to be rewritten to be both much larger and include an \[#allow\] statement.
+The expected outcome is for irrefutable patterns to be compiled to a tautology and have the if block accept it as if it was `if true { }`.
+To support this, currently you must do something roughly the following, which seems to counteract the benefit of having if-let and while-let in the spec.
+
+```rust
+#[allow(unreachable_patterns)]
+match $val {
+    $p => { $b; },
+    _ => ()
+}
+```
+
+```rust
+if let $p = $val {
+    $b
+}
+```
```

Comment:
These two adjacent code blocks are confusing, especially since the second one starts a new thought.